<a href="https://colab.research.google.com/github/vinayakghatul/Netflix_EDA/blob/main/DV_Netflix_Case.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Import basic libraries
Import Data set for Netflix

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [3]:
from google.colab import files

uploaded = files.upload()

Saving netflix.csv to netflix.csv


Data set basics - Head/Tail etc

In [4]:
import io
df = pd.read_csv(io.BytesIO(uploaded['netflix.csv']))
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [5]:
df.shape

(8807, 12)

In [6]:
#check basic datatypes
df.dtypes

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

## **Basic Analysis**
1. Un-nesting the columns
2. Handling null values

### Comma separated values in director column

In [7]:
df[df['director'].apply(lambda x: "," in str(x))]

show_id   type                                              title  \
6         s7  Movie                   My Little Pony: A New Generation   
16       s17  Movie  Europe's Most Dangerous Man: Otto Skorzeny in ...   
23       s24  Movie       Go! Go! Cory Carson: Chrissy Takes the Wheel   
30       s31  Movie                                    Ankahi Kahaniya   
68       s69  Movie                                         Schumacher   
...      ...    ...                                                ...   
8727   s8728  Movie                              When Hari Got Married   
8728   s8729  Movie                            When Two Worlds Collide   
8737   s8738  Movie                         Why Are We Getting So Fat?   
8739   s8740  Movie                 Why We Fight: The Battle of Russia   
8765   s8766  Movie                                                 XX   

                                               director  \
6                         Robert Cullen, José Luis Ucha   
16        Pedro de Echave García, Pablo Azorín Williams   
23                              Alex Woo, Stanley Moore   
30    Ashwiny Iyer Tiwari, Abhishek Chaubey, Saket C...   
68    Hanns-Bruno Kammertöns, Vanessa Nöcker, Michae...   
...                                                 ...   
8727                          Ritu Sarin, Tenzing Sonam   
8728                    Heidi Brandenburg, Mathew Orzel   
8737             Milla Harrison-Hansley, Alicky Sussman   
8739                        Frank Capra, Anatole Litvak   
8765  Jovanka Vuckovic, Annie Clark, Roxanne Benjami...   

                                                   cast  \
6     Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
16                                                  NaN   
23    Maisie Benson, Paul Killam, Kerry Gudjohnsen, ...   
30    Abhishek Banerjee, Rinku Rajguru, Delzad Hiwal...   
68                                   Michael Schumacher   
...                                                 ...   
8727                                                NaN   
8728                                                NaN   
8737                                          Giles Yeo   
8739                                                NaN   
8765  Natalie Brown, Jonathan Watton, Melanie Lynske...   

                                   country          date_added  release_year  \
6                                      NaN  September 24, 2021          2021   
16                                     NaN  September 22, 2021          2020   
23                                     NaN  September 21, 2021          2021   
30                                     NaN  September 17, 2021          2021   
68                                     NaN  September 15, 2021          2021   
...                                    ...                 ...           ...   
8727  United Kingdom, India, United States   December 25, 2016          2013   
8728   Peru, United States, United Kingdom   November 30, 2016          2016   
8737                        United Kingdom    February 1, 2019          2016   
8739                         United States      March 31, 2017          1943   
8765                 Canada, United States       June 22, 2017          2017   

     rating duration                                          listed_in  \
6        PG   91 min                           Children & Family Movies   
16    TV-MA   67 min                Documentaries, International Movies   
23     TV-Y   61 min                           Children & Family Movies   
30    TV-14  111 min   Dramas, Independent Movies, International Movies   
68    TV-14  113 min  Documentaries, International Movies, Sports Mo...   
...     ...      ...                                                ...   
8727     NR   75 min                Documentaries, International Movies   
8728  TV-14  103 min                Documentaries, International Movies   
8737  TV-14   50 min                                      Docu

### Un-nesting 'director' column
comma separated values are present in director column hence we will be un-nesting these values using title as index column

In [8]:
# Un-nesting using title and directors
constraint1 = df['director'].apply(lambda x: str(x).split(',')).tolist()
df_new1 = pd.DataFrame(constraint1, index=df['title'])
df_new1 = df_new1.stack()
df_new1 = pd.DataFrame(df_new1.reset_index())
df_new1.rename(columns={0:'Directors'},inplace=True)
df_new1.drop(['level_1'],axis=1,inplace=True)
df_new1.head()

title        Directors
0   Dick Johnson Is Dead  Kirsten Johnson
1          Blood & Water              nan
2              Ganglands  Julien Leclercq
3  Jailbirds New Orleans              nan
4           Kota Factory              nan

### Handling comma separated values for 'cast' column

In [9]:
df[df['cast'].apply(lambda x: "," in str(x))]

show_id     type                             title  \
1         s2  TV Show                     Blood & Water   
2         s3  TV Show                         Ganglands   
4         s5  TV Show                      Kota Factory   
5         s6  TV Show                     Midnight Mass   
6         s7    Movie  My Little Pony: A New Generation   
...      ...      ...                               ...   
8801   s8802    Movie                           Zinzana   
8802   s8803    Movie                            Zodiac   
8804   s8805    Movie                        Zombieland   
8805   s8806    Movie                              Zoom   
8806   s8807    Movie                            Zubaan   

                           director  \
1                               NaN   
2                   Julien Leclercq   
4                               NaN   
5                     Mike Flanagan   
6     Robert Cullen, José Luis Ucha   
...                             ...   
8801                Majid Al Ansari   
8802                  David Fincher   
8804                Ruben Fleischer   
8805                   Peter Hewitt   
8806                    Mozez Singh   

                                                   cast  \
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...   
5     Kate Siegel, Zach Gilford, Hamish Linklater, H...   
6     Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
...                                                 ...   
8801  Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                           country          date_added  release_year rating  \
1                     South Africa  September 24, 2021          2021  TV-MA   
2                              NaN  September 24, 2021          2021  TV-MA   
4                            India  September 24, 2021          2021  TV-MA   
5                              NaN  September 24, 2021          2021  TV-MA   
6                              NaN  September 24, 2021          2021     PG   
...                            ...                 ...           ...    ...   
8801  United Arab Emirates, Jordan       March 9, 2016          2015  TV-MA   
8802                 United States   November 20, 2019          2007      R   
8804                 United States    November 1, 2019          2009      R   
8805                 United States    January 11, 2020          2006     PG   
8806                         India       March 2, 2019          2015  TV-14   

       duration                                          listed_in  \
1     2 Seasons    International TV Shows, TV Dramas, TV Mysteries   
2      1 Season  Crime TV Shows, International TV Shows, TV Act...   
4     2 Seasons  International TV Shows, Romantic TV Shows, TV ...   
5      1 Season                 TV Dramas, TV Horror, TV Mysteries   
6        91 min                           Children & Family Movies   
...         ...                                                ...   
8801     96 min            Dramas, International Movies, Thrillers   
8802    158 min                     Cult Movies, Dramas, Thrillers   
8804     88 min                            Comedies, Horror Movies   
8805     88 min                 Children & Family Movies, Comedies   
8806    111 min     Dramas, International Movies, Music & Musicals   

                                            description  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
4     In a city of coaching centers known to train I...  
5     The arrival of a charismatic young priest brin...  
6     Equestria's divided. But a brigh

In [10]:
# Un-nesting using title and cast
constraint2 = df['cast'].apply(lambda x: str(x).split(',')).to_list()
df_new2 = pd.DataFrame(constraint2, index=df['title'])
df_new2 = df_new2.stack()
df_new2 = pd.DataFrame(df_new2.reset_index())
df_new2.rename(columns={0:'Actors'},inplace=True)
df_new2.drop(['level_1'], axis=1,inplace=True)
df_new2.head()

title           Actors
0  Dick Johnson Is Dead              nan
1         Blood & Water       Ama Qamata
2         Blood & Water      Khosi Ngema
3         Blood & Water    Gail Mabalane
4         Blood & Water   Thabang Molaba

### Handling nested & comma separated values for 'listed_in' column

In [11]:
# Un-nesting 'listed_in' column using title as index
constraint3 = df['listed_in'].apply(lambda x: str(x).split(',')).to_list()
df_new3 = pd.DataFrame(constraint3, index=df['title'])
df_new3 = df_new3.stack()
df_new3 = pd.DataFrame(df_new3.reset_index())
df_new3.rename(columns={0:'Genre'},inplace=True)
df_new3.drop(['level_1'],axis=1,inplace=True)
df_new3.head()

title                   Genre
0  Dick Johnson Is Dead           Documentaries
1         Blood & Water  International TV Shows
2         Blood & Water               TV Dramas
3         Blood & Water            TV Mysteries
4             Ganglands          Crime TV Shows

### Handling nested values for 'country' column

In [12]:
constraint4 = df['country'].apply(lambda x: str(x).split(',')).to_list()
df_new4 = pd.DataFrame(constraint4, index=df['title'])
df_new4 = df_new4.stack()
df_new4 = pd.DataFrame(df_new4.reset_index())
df_new4.rename(columns={0:'country'},inplace=True)
df_new4.drop(['level_1'],axis=1,inplace=True)
df_new4.head()

title        country
0   Dick Johnson Is Dead  United States
1          Blood & Water   South Africa
2              Ganglands            nan
3  Jailbirds New Orleans            nan
4           Kota Factory          India

### Merge un-nested columns on the basic of 'title' with inner join

In [13]:
#using merge in pandas to merge the data frames created in previous step

#mergeing 1 and 2
df_new5 = df_new2.merge(df_new1, on=['title'], how='inner')

#merging 5 and 3
df_new6 = df_new5.merge(df_new3, on=['title'], how='inner')

#mergeing 6 and 4
df_new = df_new6.merge(df_new4, on=['title'], how='inner')

df_new.head()

title        Actors        Directors  \
0  Dick Johnson Is Dead           nan  Kirsten Johnson   
1         Blood & Water    Ama Qamata              nan   
2         Blood & Water    Ama Qamata              nan   
3         Blood & Water    Ama Qamata              nan   
4         Blood & Water   Khosi Ngema              nan   

                    Genre        country  
0           Documentaries  United States  
1  International TV Shows   South Africa  
2               TV Dramas   South Africa  
3            TV Mysteries   South Africa  
4  International TV Shows   South Africa

### Handling 'nan' values
Replacing nan values for Actors, Directors with 'Unknown Actor' & 'Unknown Director'

In [14]:
#Actors
df_new['Actors'].replace(['nan'],['Unknown Actor'], inplace=True)
#Directors
df_new['Directors'].replace(['nan'],['Unknown Director'], inplace=True)
#Country
df_new['country'].replace(['nan'],[np.nan],inplace=True)

df_new.head()

title         Actors         Directors  \
0  Dick Johnson Is Dead  Unknown Actor   Kirsten Johnson   
1         Blood & Water     Ama Qamata  Unknown Director   
2         Blood & Water     Ama Qamata  Unknown Director   
3         Blood & Water     Ama Qamata  Unknown Director   
4         Blood & Water    Khosi Ngema  Unknown Director   

                    Genre        country  
0           Documentaries  United States  
1  International TV Shows   South Africa  
2               TV Dramas   South Africa  
3            TV Mysteries   South Africa  
4  International TV Shows   South Africa

In [15]:
#merging our unnested data with the original data
df_final=df_new.merge(df[['show_id', 'type', 'title', 'date_added',
       'release_year', 'rating', 'duration']],on=['title'],how='left')
df_final.head()

title         Actors         Directors  \
0  Dick Johnson Is Dead  Unknown Actor   Kirsten Johnson   
1         Blood & Water     Ama Qamata  Unknown Director   
2         Blood & Water     Ama Qamata  Unknown Director   
3         Blood & Water     Ama Qamata  Unknown Director   
4         Blood & Water    Khosi Ngema  Unknown Director   

                    Genre        country show_id     type          date_added  \
0           Documentaries  United States      s1    Movie  September 25, 2021   
1  International TV Shows   South Africa      s2  TV Show  September 24, 2021   
2               TV Dramas   South Africa      s2  TV Show  September 24, 2021   
3            TV Mysteries   South Africa      s2  TV Show  September 24, 2021   
4  International TV Shows   South Africa      s2  TV Show  September 24, 2021   

   release_year rating   duration  
0          2020  PG-13     90 min  
1          2021  TV-MA  2 Seasons  
2          2021  TV-MA  2 Seasons  
3          2021  TV-MA  2 Seasons  
4          2021  TV-MA  2 Seasons

### Handling Null Values


In [16]:
#count of null values
df_final.isnull().sum()

title               0
Actors              0
Directors           0
Genre               0
country         11897
show_id             0
type                0
date_added        158
release_year        0
rating             67
duration            3
dtype: int64

Duration values (3) which are null are present in rating columns as minutes from the observation which should be present in the duration column hence moving them back as follows

In [17]:
df_final[df_final['duration'].isnull()]

title      Actors   Directors   Genre  \
126582                       Louis C.K. 2017  Louis C.K.  Louis C.K.  Movies   
131648                 Louis C.K.: Hilarious  Louis C.K.  Louis C.K.  Movies   
131782  Louis C.K.: Live at the Comedy Store  Louis C.K.  Louis C.K.  Movies   

              country show_id   type          date_added  release_year  \
126582  United States   s5542  Movie       April 4, 2017          2017   
131648  United States   s5795  Movie  September 16, 2016          2010   
131782  United States   s5814  Movie     August 15, 2016          2015   

        rating duration  
126582  74 min      NaN  
131648  84 min      NaN  
131782  66 min      NaN

In [18]:
df_final.loc[df_final['duration'].isnull(), 'duration']  = df_final.loc[df_final['duration'].isnull(), 'duration'].fillna(df_final['rating'])


In [19]:
#ratings which had false values with minutes need to be updated with 'NR' - Not Rated
df_final.loc[df_final['rating'].str.contains('min', na=False), 'rating'] = 'NR'
df_final['rating'].fillna('NR', inplace=True)


In [20]:
df_final.isnull().sum()

title               0
Actors              0
Directors           0
Genre               0
country         11897
show_id             0
type                0
date_added        158
release_year        0
rating              0
duration            0
dtype: int64

In [21]:
pd.set_option('display.max_rows',None)

### Handle NaN values for 'date_added' column

In [22]:
df_final[df_final['date_added'].isnull()].head()

title            Actors  \
136940  A Young Doctor's Notebook and Other Stories  Daniel Radcliffe   
136941  A Young Doctor's Notebook and Other Stories  Daniel Radcliffe   
136942  A Young Doctor's Notebook and Other Stories  Daniel Radcliffe   
136943  A Young Doctor's Notebook and Other Stories          Jon Hamm   
136944  A Young Doctor's Notebook and Other Stories          Jon Hamm   

               Directors             Genre         country show_id     type  \
136940  Unknown Director  British TV Shows  United Kingdom   s6067  TV Show   
136941  Unknown Director       TV Comedies  United Kingdom   s6067  TV Show   
136942  Unknown Director         TV Dramas  United Kingdom   s6067  TV Show   
136943  Unknown Director  British TV Shows  United Kingdom   s6067  TV Show   
136944  Unknown Director       TV Comedies  United Kingdom   s6067  TV Show   

       date_added  release_year rating   duration  
136940        NaN          2013  TV-MA  2 Seasons  
136941        NaN          2013  TV-MA  2 Seasons  
136942        NaN          2013  TV-MA  2 Seasons  
136943        NaN          2013  TV-MA  2 Seasons  
136944        NaN          2013  TV-MA  2 Seasons

Imputation for 'date_added' with relevant value by referencing release_year column
Steps:
1. Calculate the mode for corresponding release year
2. If 'date_added' is 'NaN' then using release year fetch mode value and replace it in 'date_added' column

In [23]:
for year in df_final[df_final['date_added'].isnull()]['release_year'].unique():
  imp = df_final[df_final['release_year']==year]['date_added'].mode().values[0]
  df_final.loc[df_final['release_year'] == year, 'date_added'] = df_final.loc[df_final['release_year'] == year,'date_added'].fillna(imp)


### Imputation for Country on the basic of Directors/Actors
For certain title of 'country' value is missing then it can be imputed with it's director's mode value in one iteration and it's actor's mode value in the second iteration.
Remaining unknown values can be replaced with 'Unknown Country'

In [24]:
for director in df_final[df_final['country'].isnull()]['Directors'].unique():
  if director in df_final[~df_final['country'].isnull()]['Directors'].unique():
    imp = df_final[df_final['Directors']==director]['country'].mode().values[0]
    df_final.loc[df_final['Directors']==director, 'country'] = df_final.loc[df_final['Directors']==director, 'country'].fillna(imp)

In [25]:
len(df_final[df_final['country'].isnull()])

4673

Still there is significant number of null values even after using 'Directors' column for imputation.
We will use 'Actors' column for further reducing this count.

In [26]:
for actor in df_final[df_final['country'].isnull()]['Actors'].unique():
  if actor in df_final[~df_final['country'].isnull()]['Actors'].unique():
    imp = df_final[df_final['Actors']==actor]['country'].mode().values[0]
    df_final.loc[df_final['Actors']==actor,'country'] = df_final.loc[df_final['Actors']==actor,'country'].fillna(imp)

In [27]:
len(df_final[df_final['country'].isnull()])

2455

Still there are 2455 rows with country value as 'NaN'.
These values can be replaced with 'Unknown Country'

In [28]:
df_final['country'].fillna('Unknown Country', inplace=True)
len(df_final[df_final['country'].isnull()])

0

In [29]:
#Final null values tally for the dataset
df_final.isnull().sum()

title           0
Actors          0
Directors       0
Genre           0
country         0
show_id         0
type            0
date_added      0
release_year    0
rating          0
duration        0
dtype: int64

In [33]:
df_final.head()

title         Actors         Directors  \
0  Dick Johnson Is Dead  Unknown Actor   Kirsten Johnson   
1         Blood & Water     Ama Qamata  Unknown Director   
2         Blood & Water     Ama Qamata  Unknown Director   
3         Blood & Water     Ama Qamata  Unknown Director   
4         Blood & Water    Khosi Ngema  Unknown Director   

                    Genre        country show_id     type          date_added  \
0           Documentaries  United States      s1    Movie  September 25, 2021   
1  International TV Shows   South Africa      s2  TV Show  September 24, 2021   
2               TV Dramas   South Africa      s2  TV Show  September 24, 2021   
3            TV Mysteries   South Africa      s2  TV Show  September 24, 2021   
4  International TV Shows   South Africa      s2  TV Show  September 24, 2021   

   release_year rating   duration  
0          2020  PG-13     90 min  
1          2021  TV-MA  2 Seasons  
2          2021  TV-MA  2 Seasons  
3          2021  TV-MA  2 Seasons  
4          2021  TV-MA  2 Seasons

# Answering the questions asked in the case


###

Following fields have comma separated values that can cause issues:


*   **director** - movies/shows with multiple director
*   **cast** -  multiple values for each movie/show
*   **country** - multiple values for each movie/show
*   **listed_in** - listed in multiple genres

These are the columns which required un-nesting

### Handling missing/null/NaN values
Best Way - Multiple imputation (Missing data being replaced with something closely relevant as per the column)

Ex. For directors - 'Not Available' can be used. Country wise best/popular director can be used

For director - 'Louis C.K.' rating values are misplaced with duration and duration values are NaN

In [ ]:
df[df['director'].apply(lambda x: x=='Louis C.K.')]

show_id   type                                 title    director  \
5541   s5542  Movie                       Louis C.K. 2017  Louis C.K.   
5794   s5795  Movie                 Louis C.K.: Hilarious  Louis C.K.   
5813   s5814  Movie  Louis C.K.: Live at the Comedy Store  Louis C.K.   

            cast        country          date_added  release_year  rating  \
5541  Louis C.K.  United States       April 4, 2017          2017  74 min   
5794  Louis C.K.  United States  September 16, 2016          2010  84 min   
5813  Louis C.K.  United States     August 15, 2016          2015  66 min   

     duration listed_in                                        description  
5541      NaN    Movies  Louis C.K. muses on religion, eternal love, gi...  
5794      NaN    Movies  Emmy-winning comedy writer Louis C.K. brings h...  
5813      NaN    Movies  The comic puts his trademark hilarious/thought...

### null values count

In [ ]:
df.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [ ]:
df[df['rating'].isnull()]

show_id     type                                              title  \
5989   s5990    Movie  13TH: A Conversation with Oprah Winfrey & Ava ...   
6827   s6828  TV Show                  Gargantia on the Verdurous Planet   
7312   s7313  TV Show                                       Little Lunch   
7537   s7538    Movie                               My Honor Was Loyalty   

             director                                               cast  \
5989              NaN                        Oprah Winfrey, Ava DuVernay   
6827              NaN  Kaito Ishikawa, Hisako Kanemoto, Ai Kayano, Ka...   
7312              NaN  Flynn Curry, Olivia Deeble, Madison Lu, Oisín ...   
7537  Alessandro Pepe  Leone Frisa, Paolo Vaccarino, Francesco Miglio...   

        country        date_added  release_year rating  duration  \
5989        NaN  January 26, 2017          2017    NaN    37 min   
6827      Japan  December 1, 2016          2013    NaN  1 Season   
7312  Australia  February 1, 2018          2015    NaN  1 Season   
7537      Italy     March 1, 2017          2015    NaN   115 min   

                                 listed_in  \
5989                                Movies   
6827  Anime Series, International TV Shows   
7312                 Kids' TV, TV Comedies   
7537                                Dramas   

                                            description  
5989  Oprah Winfrey sits down with director Ava DuVe...  
6827  After falling through a wormhole, a space-dwel...  
7312  Adopting a child's perspective, this show take...  
7537  Amid the chaos and horror of World War II, a c...